samples: nrf_desktop: Add PCA20037 (keyboard)

pdunaj opened the pull request:
Jira:DESK-483

pdunaj commented:
On Android everything (that is implemented) works fine. I am able to type without any problems.
On Windows 7 (Intel driver) system blue screened when I tested (I modified the device information since then as there was leftover from mouse). I will check it again after I finish creating PR.

pdunaj commented:
Modifiers are not implemented yet as hid state has no support for them at the moment (i.e. no way to differentiate them from any other key). Will work on that under separate story.

LEDs are not used yet as our project at the moment require at least two of them while keyboard has only one. @jhn-nordic we need to think how to show various things using this single diod (connection status, battery status, numlock, caps lock) or decide what will be shown there (e.g. connection status only).

Media keys are not handled correctly yet. I think we will need to modify the report slightly and add them in the same report as other keys. (at the moment we have a special mplayer section that goes separately but will only cause problems if kept this way).

pdunaj commented:
CI fails at
```
Exception: Image size (0x3241e) + trailer (0x620) exceeds requested size 0x29000
```

This is 832 so it has less space on flash and we are not fitting. Will also check if something can be disabled or maybe we should simply disable some build variants for now.

SebastianBoe commented:
I assume you have considered optimizing for size instead of speed?

SebastianBoe commented on samples/nrf_desktop/configuration/nrf52_pca20037/buttons_def.h:
```diff
@@ -0,0 +1,45 @@
+
+#include "buttons.h"
+
+/* This file must be included only once */
```
This is not OK.

Please place col and row in a C file, and have users reference them by having them include a header file that declares them (with extern).

pdunaj replied:
Pros and cos were considered and this layout was selected for the project. I appreciate your opinion but I don't agree here.
If you feel this is not OK please file a bug under DESK and we can discuss it there. This is not the right place.

rakons replied:
We could possibly remove this problem by... the usage of Xmacros <-:[

SebastianBoe replied:
I assume there is a good reason to go against the norm of how C sources and header files are used.

Please document this somewhere and a reference to it from here.

pdunaj replied:
Hi @SebastianBoe , sure. This is a good idea.

@jhn-nordic , @Qbicz , @MarekPieta This may be a good chance to start creating documentation for the project configuration. At the same time we could check how intuitive is this part is. Created https://projecttools.nordicsemi.no/jira/browse/DESK-535 to handle this.

SebastianBoe replied:
Please rephrase this:

> This file must be included only once

which sounds like requirements documentation for something the user must take care of. To something like:

This struct enforces that this header file is only included once in the build, as including it twice would trigger a multiple definition error at link time.

SebastianBoe replied:
With this mechanism, what happens when you have two C files that need to access the information in row?

pdunaj replied:
@SebastianBoe , this is configuration file specific to a single module. The current architecture of the app will make it work as we have one file per module (this is a design choice). It is in the form of a header file is to allow optimizations of some loops in modules. We could change the module source to work around it but we would have to pay with additional maintenance costs. Using header file gives us the ability to keep the configuration in separate file, keep the maintenance costs intact and being able to optimize for various cases. The cons is that it is non standard way header files are used and this is the price we decided to pay. I think you are right we should add some comment about that in the header to make sure it is well understood that these files are meant for one module only.

SebastianBoe replied:
This is OK. I suspect that eventually, the limitation of the configuration only being accessible from a single module will cause a problem.

But until it does, I see no harm in this as long it is clearly documented what is going on here for future reviewers.

pdunaj replied:
Thanks

pdunaj replied:
Comments added.

pdunaj replied:
Fixed

SebastianBoe commented on samples/nrf_desktop/configuration/nrf52_pca20037/flash.overlay:
```diff
@@ -0,0 +1 @@
+/* Currently no dedicated partition sizes. */
```
AFAIK flash.overlay is not used any more.

pdunaj replied:
leftover after rebase. thanks

SebastianBoe commented on samples/nrf_desktop/configuration/nrf52_pca20037/hid_keymap_def.h:
```diff
@@ -0,0 +1,124 @@
+
+#include "hid_keymap.h"
+
+/* This file must be included only once */
```
Same comment here as for row and col.

Qbicz commented on samples/nrf_desktop/configuration/nrf52_pca20037/app_ZReleaseMCUBoot.conf:
```diff
@@ -0,0 +1,125 @@
+CONFIG_BT_AUTO_PHY_UPDATE=n
+
+CONFIG_BT_DEVICE_NAME="Keyboard nRF52 Desktop"
+CONFIG_BT_DEVICE_APPEARANCE=962
```
Keyboard have appearance 961 (0x3C1) 
https://www.bluetooth.org/DocMan/handlers/DownloadDoc.ashx?doc_id=429632

pdunaj replied:
missed that, thanks

pdunaj replied:
Fixed

pdunaj commented on samples/nrf_desktop/configuration/nrf52_pca20037/led_state_def.h:
```diff
@@ -0,0 +1,71 @@
+/* Map function to LED ID */
+static const u8_t led_map[LED_ID_COUNT] = {
+	[LED_ID_SYSTEM_STATE] = 0,
+	[LED_ID_PEER_STATE] = 1
```
I ignored this file for now as FW configuration does not match HW. This will have to be fixed under separate PR. We also need to discuss how to show various things with only 1 led.

pdunaj commented on samples/nrf_desktop/configuration/nrf52_pca20037/port_state_def.h:
```diff
@@ -0,0 +1,41 @@
+static const struct pin_state port0_on[] = {
+
+#if !defined(CONFIG_DESKTOP_LED_ENABLE)
+	{30, 0}, /* LED red */
```
true but I want to turn them on ;)

pdunaj replied:
I added part to turn them off when device go to sleep

pdunaj commented on samples/nrf_desktop/configuration/nrf52_pca20037/port_state_def.h:
```diff
@@ -0,0 +1,41 @@
+#endif
+};
+
+static const struct pin_state port0_off[] = {
```
I turn off leds now. Fixed

pdunaj commented on samples/nrf_desktop/README.rst:
```diff
@@ -20,6 +20,7 @@ Requirements
 Supported boards:
 * PCA10056 (nRF52840 DK) - example where the mouse is emulated using DK buttons
 * PCA20041 - nRF Desktop mouse reference design
+* PCA20037 - nRF Desktop keyboard reference design
```
Added

pdunaj commented on samples/nrf_desktop/configuration/nrf52_pca20037/hid_keymap_def.h:
```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright (c) 2018 Nordic Semiconductor ASA
```
Fixed

pdunaj commented on samples/nrf_desktop/configuration/nrf52_pca20037/led_state_def.h:
```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright (c) 2018 Nordic Semiconductor ASA
```
I remove this file until we will define how to configure them.

pdunaj commented on samples/nrf_desktop/configuration/nrf52_pca20037/buttons_def.h:
```diff
@@ -0,0 +1,45 @@
+	{ .port = 0, .pin = 14 },
+	{ .port = 0, .pin = 12 },
+	{ .port = 0, .pin = 10 },
+	{ .port = 0, .pin = 2  },
```
thanks! Fixed

pdunaj commented:
I have fixed report so keyboard works now on Windows. Also added support for modifiers.

I still need to fix building for bootloader configuration. Will do this under this PR.
We need to decide what to do with leds. (we have a story in jira)
I see that sometimes 'e' is repeated multiple times when actually pressed only once (have not seen this for any other key so far). Will create a bug for that after PR is merged and investigate.
media keys are not handled yet (will create new story)
function key is not handled (we have story in jira).

Besides that it seems basic functionality works as expected :) this comment was actually written using our keyboard :)

Qbicz commented on samples/nrf_desktop/src/modules/hid_state.c:
```diff
@@ -526,15 +526,23 @@ static void send_report_keyboard(void)
 
 			if (item.value) {
-				event->keys[cnt] = item.usage_id;
+				if (item.usage_id < 0x66) {
```
What is 0x66, 0xE0, 0xE7? Can you name them or at least comment?

pdunaj replied:
Fixed

pdunaj commented:
The PR merging is blocked and it seems it won't be possible to have waiver for this release. Due to that I will close this PR and will open new PRs on dev branch.

@carlescufi can we create a project dev branch on this repo? I would like to avoid creating that on personal fork.

@wiminordic can you tell if we can have CI on dev branch?

wiminordic commented:
CI should run on a branch as long as there is Jenkins file present (you would have to remove it to make CI not run).

pdunaj commented:
@wiminordic , thanks for confirming things. I will check with @carlescufi when he is available and create a dev branch for us. Will let you know when this is done.

pdunaj commented:
I pushed the changes. When tests are green I will move PR to dev branch so we can merge it and continue work there.

pdunaj commented:
It's green. I am creating a dev branch now...

pdunaj commented:
@jhn-nordic , @Qbicz , @MarekPieta , @carlescufi 
https://github.com/NordicPlayground/fw-nrfconnect-nrf/tree/nrf_desktop_dev

pdunaj commented:
This change is merged to nrf_desktop_dev branch and will be moved back to master after the release is complete. I am closing this PR.
